<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from robot.urdf.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="zmr_robot">
  <!-- To get a URDF file, run: `ros2 run xacro xacro robot.urdf.xacro > robot.urdf` -->
  <!-- Material colors -->
  <material name="blue_body">
    <color rgba="0.1 0.1 1.0 0.8"/>
  </material>
  <material name="black_laser">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="black_wheel">
    <color rgba="0.0 0.0 0.0 0.8"/>
  </material>
  <material name="black_caster">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <!-- [kg] -->
  <!-- Gazebo - differential drive -->
  <gazebo>
    <plugin filename="libgazebo_ros_diff_drive.so" name="diff_drive">
      <ros>
        <namespace>zmr_1</namespace>
        <remapping>cmd_vel:=cmd_vel</remapping>
        <remapping>odom:=odom</remapping>
      </ros>
      <update_rate>30</update_rate>
      <!-- wheels -->
      <left_joint>zmr_1/left_wheel_joint</left_joint>
      <right_joint>zmr_1/right_wheel_joint</right_joint>
      <!-- kinematics -->
      <wheel_separation>0.2</wheel_separation>
      <!-- 1.0-(-1.0) -->
      <wheel_diameter>0.065</wheel_diameter>
      <!-- ~0.032*2 -->
      <!-- limits -->
      <max_wheel_torque>20</max_wheel_torque>
      <max_wheel_acceleration>1.0</max_wheel_acceleration>
      <!-- output -->
      <publish_odom>true</publish_odom>
      <publish_odom_tf>true</publish_odom_tf>
      <publish_wheel_tf>false</publish_wheel_tf>
      <odometry_frame>zmr_1/odom</odometry_frame>
      <robot_base_frame>zmr_1/base_footprint</robot_base_frame>
    </plugin>
    <plugin filename="libgazebo_ros_joint_state_publisher.so" name="zmr_joint_state">
      <ros>
        <namespace>zmr_1</namespace>
        <remapping>~/out:=joint_states</remapping>
      </ros>
      <update_rate>30</update_rate>
      <joint_name>zmr_1/left_wheel_joint</joint_name>
      <joint_name>zmr_1/right_wheel_joint</joint_name>
    </plugin>
  </gazebo>
  <!-- base link -->
  <link name="zmr_1/base_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder length="0.12" radius="0.1"/>
      </geometry>
      <material name="blue_body"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <cylinder length="0.12" radius="0.1"/>
      </geometry>
      <material name="blue_body"/>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <inertia ixx="0.0007400000000000002" ixy="0" ixz="0" iyy="0.0007400000000000002" iyz="0" izz="0.0010000000000000002"/>
    </inertial>
  </link>
  <!-- laser link -->
  <link name="zmr_1/laser_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.02"/>
      </geometry>
      <material name="black_laser"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.02" radius="0.02"/>
      </geometry>
      <material name="black_laser"/>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="1.3333333333333332e-05" ixy="0" ixz="0" iyy="1.3333333333333332e-05" iyz="0" izz="2e-05"/>
    </inertial>
  </link>
  <!-- laser joint -->
  <joint name="zmr_1/laser_joint" type="fixed">
    <parent link="zmr_1/base_link"/>
    <child link="zmr_1/laser_link"/>
    <origin xyz="0 0 0.075"/>
  </joint>
  <!-- imu link -->
  <link name="zmr_1/imu_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="6.6666666666666675e-06" ixy="0" ixz="0" iyy="6.6666666666666675e-06" iyz="0" izz="6.6666666666666675e-06"/>
    </inertial>
  </link>
  <!-- imu joint -->
  <joint name="zmr_1/imu_joint" type="fixed">
    <parent link="zmr_1/base_link"/>
    <child link="zmr_1/imu_link"/>
    <origin xyz="0 0 0.02"/>
  </joint>
  <!-- right wheel link -->
  <link name="zmr_1/right_wheel_link">
    <visual>
      <origin rpy="1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.032"/>
      </geometry>
      <material name="black_wheel"/>
    </visual>
    <collision>
      <origin rpy="1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.032"/>
      </geometry>
      <material name="black_wheel"/>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <inertia ixx="7.786666666666667e-05" ixy="0" ixz="0" iyy="7.786666666666667e-05" iyz="0" izz="0.00010240000000000001"/>
    </inertial>
  </link>
  <!-- right wheel joint -->
  <joint name="zmr_1/right_wheel_joint" type="continuous">
    <parent link="zmr_1/base_link"/>
    <child link="zmr_1/right_wheel_link"/>
    <origin xyz="-0.02 -0.10 -0.06"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- left wheel link -->
  <link name="zmr_1/left_wheel_link">
    <visual>
      <origin rpy="1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.032"/>
      </geometry>
      <material name="black_wheel"/>
    </visual>
    <collision>
      <origin rpy="1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.04" radius="0.032"/>
      </geometry>
      <material name="black_wheel"/>
    </collision>
    <inertial>
      <mass value="0.2"/>
      <inertia ixx="7.786666666666667e-05" ixy="0" ixz="0" iyy="7.786666666666667e-05" iyz="0" izz="0.00010240000000000001"/>
    </inertial>
  </link>
  <!-- left wheel joint -->
  <joint name="zmr_1/left_wheel_joint" type="continuous">
    <parent link="zmr_1/base_link"/>
    <child link="zmr_1/left_wheel_link"/>
    <origin xyz="-0.02 0.10 -0.06"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- caster link -->
  <link name="zmr_1/caster_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <sphere radius="0.016"/>
      </geometry>
      <material name="black_caster"/>
    </visual>
    <collision>
      <origin rpy="1.57079 0 0" xyz="0 0 0"/>
      <geometry>
        <sphere radius="0.016"/>
      </geometry>
      <material name="black_caster"/>
    </collision>
    <inertial>
      <mass value="0.02"/>
      <inertia ixx="2.048e-06" ixy="0" ixz="0" iyy="2.048e-06" iyz="0" izz="2.048e-06"/>
    </inertial>
  </link>
  <!-- caster joint -->
  <joint name="zmr_1/caster_joint" type="fixed">
    <parent link="zmr_1/base_link"/>
    <child link="zmr_1/caster_link"/>
    <origin xyz="0.06 0.0 -0.076"/>
  </joint>
  <!-- Gazebo - caster joint -->
  <gazebo reference="zmr_1/caster_link">
    <material>Gazebo/Black</material>
    <mu1 value="0.0"/>
    <mu2 value="0.0"/>
    <kp value="1000000.0"/>
    <kd value="10.0"/>
    <!-- <fdir1 value="0 0 1"/> -->
  </gazebo>
  <!-- Gazebo - imu link -->
  <gazebo reference="zmr_1/imu_link">
    <sensor name="imu_sensor" type="imu">
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <ros>
          <namespace>zmr_1</namespace>
          <remapping>~/out:=imu</remapping>
        </ros>
        <initial_orientation_as_reference>false</initial_orientation_as_reference>
      </plugin>
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <imu>
        <angular_velocity>
          <x>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
              <bias_mean>0.0000075</bias_mean>
              <bias_stddev>0.0000008</bias_stddev>
            </noise>
          </x>
          <y>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
              <bias_mean>0.0000075</bias_mean>
              <bias_stddev>0.0000008</bias_stddev>
            </noise>
          </y>
          <z>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>2e-4</stddev>
              <bias_mean>0.0000075</bias_mean>
              <bias_stddev>0.0000008</bias_stddev>
            </noise>
          </z>
        </angular_velocity>
        <linear_acceleration>
          <x>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
              <bias_mean>0.1</bias_mean>
              <bias_stddev>0.001</bias_stddev>
            </noise>
          </x>
          <y>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
              <bias_mean>0.1</bias_mean>
              <bias_stddev>0.001</bias_stddev>
            </noise>
          </y>
          <z>
            <noise type="gaussian">
              <mean>0.0</mean>
              <stddev>1.7e-2</stddev>
              <bias_mean>0.1</bias_mean>
              <bias_stddev>0.001</bias_stddev>
            </noise>
          </z>
        </linear_acceleration>
      </imu>
    </sensor>
  </gazebo>
  <!-- Gazebo - laser link -->
  <gazebo reference="zmr_1/laser_link">
    <sensor name="laser_sensor" type="ray">
      <always_on>true</always_on>
      <visualize>true</visualize>
      <material>Gazebo/Black</material>
      <update_rate>5</update_rate>
      <pose>0 0 0.075 0 0 0</pose>
      <ray>
        <scan>
          <horizontal>
            <samples>360</samples>
            <resolution>1.000000</resolution>
            <min_angle>0.000000</min_angle>
            <max_angle>6.280000</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.120000</min>
          <max>3.5</max>
          <resolution>0.015000</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.01</stddev>
        </noise>
      </ray>
      <plugin filename="libgazebo_ros_ray_sensor.so" name="laserscan">
        <ros>
          <namespace>zmr_1</namespace>
          <remapping>~/out:=scan</remapping>
        </ros>
        <output_type>sensor_msgs/LaserScan</output_type>
        <frame_name>zmr_1/laser_link</frame_name>
      </plugin>
    </sensor>
  </gazebo>
  <!-- Robot Footprint link and joint -->
  <link name="zmr_1/base_footprint"/>
  <joint name="zmr_1/base_joint" type="fixed">
    <parent link="zmr_1/base_footprint"/>
    <child link="zmr_1/base_link"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 0.092"/>
    <!-- 0.092=0.016*2+0.12/2, caster and main body -->
  </joint>
</robot>
